### PR TITLE
Add ui-spacer

### LINKED ISSUE #1415: UI Element: Spacer
### Description

In the original Dashboard, it was possible to add "Spacers", which would just leave blank white space in the application It meant you had more control over alignment of your elements. 

This would be a config node, and addable via the Dashboard 2.0 side panel.

In the UI itself, it would not render any content within the widget, but just leave it as a blank space. The space should have a width/height option, and nothing else.

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### Requested By:

https://app-eu1.hubspot.com/help-desk/26586079/view/233410279/ticket/8321776358/thread/4745614815#live-chat

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #1415

## Description

* Adds a ui-spacer node
* updates ui_base to support adding, configuring, sorting of new spacer

### NOTES:
* Depending on whether this or https://github.com/FlowFuse/node-red-dashboard/pull/1369 is merged first, one of them will require some retro work
* Depending on whether this or https://github.com/FlowFuse/node-red-dashboard/pull/1428 is merged first, one of them will require some retro work

### Known Issues & shortcomings
* Spacers are added to global flow as config nodes. That means they are not exported with a group. The only way to export a group with spacers is to export the entire node-red instance flows.
  * A potential workaround would be to add an `Export Group` action to the sidebar that generates the flow json for the entire group and all widgets within. something that would be useful for sharing purposes and not easily possible today (since widgets can live on separate flow tabs)
  * This is the same for DB1 (kind of: they use a hack to add spacers to the flow-tab scope (by setting `z` prop) meaning you can export spacers with the group IF the user exports the whole flow-tab
     * the issue here is widgets are not locked to a single flow-tab so this is pontless and messy and easily gets out of step since the z: propery applied to DB1 spacers is the Active Tab. i.e. it is super simple to get spacers in a group belonging to mixed flow-tabs.
* Editing a spacer from the `config` sizebar does not work. The node edit panels `group` config selector is not initialised. This is the same as DB1

### TODO:
- [x] tests
- [x] docs

## Related Issue(s)

#1415

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
Demo

```
[{"id":"07f7c6534787ceb5","type":"tab","label":"Flow 1","disabled":false,"info":"","env":[]},{"id":"637aa00d7b3575b5","type":"ui-group","name":"Group 1","page":"5ae5d5af973be1de","width":"12","height":"1","order":1,"showTitle":true,"className":"group1","visible":"true","disabled":"false","groupType":"default"},{"id":"9d33c3bb35555d6e","type":"ui-theme","name":"Default Theme","colors":{"surface":"#ffffff","primary":"#0094ce","bgPage":"#eeeeee","groupBg":"#ffffff","groupOutline":"#cccccc"},"sizes":{"density":"default","pagePadding":"12px","groupGap":"12px","groupBorderRadius":"4px","widgetGap":"12px"}},{"id":"3ce5edbdfd346f1d","type":"ui-spacer","group":"637aa00d7b3575b5","name":"spacer","tooltip":"","order":5,"width":"1","height":"3","className":"blue spacer-text"},{"id":"d4f767180862412a","type":"ui-spacer","group":"637aa00d7b3575b5","name":"spacer","tooltip":"","order":3,"width":"3","height":"1","className":"green spacer-text spacer2"},{"id":"3272ba16e1424992","type":"ui-spacer","group":"637aa00d7b3575b5","name":"spacer","tooltip":"","order":1,"width":"6","height":"6","className":"yellow spacer-text spacer1"},{"id":"98f4d4b0289e4187","type":"ui_base","theme":{"name":"theme-light","lightTheme":{"default":"#0094CE","baseColor":"#0094CE","baseFont":"-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif","edited":true,"reset":false},"darkTheme":{"default":"#097479","baseColor":"#097479","baseFont":"-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif","edited":false},"customTheme":{"name":"Untitled Theme 1","default":"#4B7930","baseColor":"#4B7930","baseFont":"-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif"},"themeState":{"base-color":{"default":"#0094CE","value":"#0094CE","edited":false},"page-titlebar-backgroundColor":{"value":"#0094CE","edited":false},"page-backgroundColor":{"value":"#fafafa","edited":false},"page-sidebar-backgroundColor":{"value":"#ffffff","edited":false},"group-textColor":{"value":"#1bbfff","edited":false},"group-borderColor":{"value":"#ffffff","edited":false},"group-backgroundColor":{"value":"#ffffff","edited":false},"widget-textColor":{"value":"#111111","edited":false},"widget-backgroundColor":{"value":"#0094ce","edited":false},"widget-borderColor":{"value":"#ffffff","edited":false},"base-font":{"value":"-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica Neue,sans-serif"}},"angularTheme":{"primary":"indigo","accents":"blue","warn":"red","background":"grey","palette":"light"}},"site":{"name":"Node-RED Dashboard","hideToolbar":"false","allowSwipe":"false","lockMenu":"false","allowTempTheme":"true","dateFormat":"DD/MM/YYYY","sizes":{"sx":48,"sy":48,"gx":6,"gy":6,"cx":6,"cy":6,"px":0,"py":0}}},{"id":"5ae5d5af973be1de","type":"ui-page","name":"Page 1","ui":"aec349d2c1fefdec","path":"/page1","icon":"home","layout":"grid","theme":"9d33c3bb35555d6e","breakpoints":[{"name":"Default","px":"0","cols":"3"},{"name":"Tablet","px":"576","cols":"6"},{"name":"Small Desktop","px":"768","cols":"9"},{"name":"Desktop","px":"1024","cols":"12"}],"order":1,"className":"","visible":true,"disabled":false},{"id":"aec349d2c1fefdec","type":"ui-base","name":"My Dashboard","path":"/dashboard","appIcon":"","includeClientData":true,"acceptsClientConfig":["ui-notification","ui-control"],"showPathInSidebar":false,"showPageTitle":true,"navigationStyle":"icon","titleBarStyle":"default"},{"id":"17bb9916321dc34e","type":"ui-spacer","group":"637aa00d7b3575b5","name":"spacer","tooltip":"","order":6,"width":"1","height":"1","className":""},{"id":"4573fbbd62f5e7cb","type":"ui-text-input","z":"07f7c6534787ceb5","group":"637aa00d7b3575b5","name":"text 3x1","label":"text","order":4,"width":"3","height":"1","topic":"topic","topicType":"msg","mode":"text","tooltip":"","delay":300,"passthru":true,"sendOnDelay":false,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"sendOnClear":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":120,"y":80,"wires":[[]]},{"id":"f768432185d6b32b","type":"ui-button","z":"07f7c6534787ceb5","group":"637aa00d7b3575b5","name":"but 3x3","label":"button","order":2,"width":"3","height":"3","emulateClick":false,"tooltip":"","color":"","bgcolor":"","className":"green","icon":"","iconPosition":"left","payload":"","payloadType":"str","topic":"topic","topicType":"msg","buttonColor":"","textColor":"","iconColor":"","enableClick":true,"enablePointerdown":false,"pointerdownPayload":"","pointerdownPayloadType":"str","enablePointerup":false,"pointerupPayload":"","pointerupPayloadType":"str","x":140,"y":140,"wires":[[]]},{"id":"8a97de1bc39b07dc","type":"ui-template","z":"07f7c6534787ceb5","group":"","page":"","ui":"aec349d2c1fefdec","name":"","order":0,"width":0,"height":0,"head":"","format":"/**** Temp styles to visualise the spacers which are normally empty ****/\n.nrdb-ui-spacer {\n  background-color: lightgrey;\n}\n.nrdb-ui-spacer.yellow {\n    background-color:gold;\n}\n.nrdb-ui-spacer.green {\n    background-color: lime;\n}\n.nrdb-ui-spacer.blue {\n    background-color: cyan;\n}\n.spacer-text::before {\n  content: \"This is a spacer\";\n  font-size: 18px;\n  color: darkblue;\n} \n.spacer-text.spacer1::before {\n  content: \"This is spacer 1\";\n}\n.spacer-text.spacer2::before {\n  content: \"This is spacer 2\";\n}\n\n/**** hide spacers based on screen size ****/\n\n/* Hide spacer 1 on < lage screen */\n@media only screen and (max-width: 1024px) {\n    .spacer1 {\n        display: none !important;\n    }\n}\n\n/* Hide spacer 2 on < med screen */\n@media only screen and (max-width: 768px) {\n    .spacer2 {\n        display: none !important;\n    }\n}\n\n/* Hide all spacers on small screen */\n@media only screen and (max-width: 576px) {\n    .nrdb-ui-spacer {\n        display: none !important;\n    }\n}","storeOutMessages":true,"passthru":true,"resendOnRefresh":true,"templateScope":"site:style","className":"","x":220,"y":240,"wires":[[]]}]
```

chrome_wqaz6NkJQN

**Comment by joepavitt** [comment]
> Editing a spacer from the config sizebar does not work. The node edit panels group config selector is not initialised.

Is that an underlying NR issue?

**Comment by Steve-Mcl** [comment]
updated to use the newer node creation pattern established in #1428
updated to remove widget width invalidation in line with changes made to support wysiwyg #1369

Ready for review.

**Comment by Steve-Mcl** [comment]
> > Editing a spacer from the config sizebar does not work. The node edit panels group config selector is not initialised.
> 
> Is that an underlying NR issue?

Not entirely sure tbh. DB1 has the same issue. It is (I suspect) due to abusing the idea of a config node to be a widget. But it _might_ be since the config was altered to have the separate `+` button


Will need to do some checks with NR 3.0.2 - but in short, i suspect this is either an "always been this way" or "not a DB2 issue".

**Comment by Steve-Mcl** [comment]
> [Remove odd  that was adding an extra row to every widget](https://github.com/FlowFuse/node-red-dashboard/pull/1430/commits/07f379789456e179d5984f523249eef0290a7739)

Bah - i knew there was one last issue - well spotted joe.

**Comment by joepavitt** [comment]
Added styling to the `ui-spacer` in the Dashboard when the page is in "Edit Mode"

Screenshot 2024-11-04 at 16 43 18

**Comment by Steve-Mcl** [comment]
> Added styling to the `ui-spacer` in the Dashboard when the page is in "Edit Mode"
> Screenshot 2024-11-04 at 16 43 18

Do we want to update the text shown in the sidebar to match the edit mode placeholder bracketed "Spacer (XxY)" format?

**Comment by joepavitt** [comment]
Yeah, sensible actually

**Review by joepavitt — APPROVED** [review]

**Comment by joepavitt** [comment]
@Steve-Mcl approved subject to that minor change, feel free to merge once you're done with it

File: nodes/widgets/ui_spacer.html [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                },
+                height: { value: 1 },
+                className: { value: '' }
+            },
+            inputs: 0,
+            outputs: 0,
+            hasUsers: false,
+            icon: 'font-awesome/fa-arrows-h',
+            paletteLabel: 'spacer',
+            label: function () { return `${this.name || 'Spacer'} ${this.width} x ${this.height}` },
```
Comment (Steve-Mcl) [inline nodes/widgets/ui_spacer.html:30]:
```suggestion
            label: function () { return `${this.name || 'spacer'} (${this.width}x${this.height})` },
```

**Review by Steve-Mcl — COMMENTED** [review]
updated sidebar naming to **"Name (xxY)"**

**Comment by Steve-Mcl** [comment]
will merge after tests go green

**Pull request merged**